### PR TITLE
feat(refactor): move to support different backends for metric and group storage

### PR DESCRIPTION
**Pull request by aaronkvanmeerten**
- update RedLock and redis
- interfaces for MetricsStore and InstanceStore
- move redis-specific code to its own file
- initial prometheus support for metrics
- update tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,13 +23,15 @@
...
+    "ioredis": "^5.4.1",
     "jsonwebtoken": "^9.0.2",
     "node-cache": "^5.1.2",
     "oci-sdk": "^1.5.2",
     "prom-client": "^15.0.0",
+    "prometheus-query": "^3.4.1",
+    "prometheus-remote-write": "^0.4.1",
     "redis": "^3.1.2",
-    "redlock": "^4.2.0",
+    "redlock": "^5.0.0-beta.2",
```
Comment (saghul) [inline package.json:34]: why use a beta release? anything specific we need from it?

File: src/app.ts [hunk 2]
```diff
@@ -48,28 +51,68 @@ const redisOptions = <RedisOptions>{
...
+
+switch (config.MetricsStoreProvider) {
+    case 'prometheus':
+        metricsStore = new PrometheusClient({
+            logger,
+            endpoint: config.PrometheusURL,
+        });
+        break;
+    default:
+    case 'redis':
```
Comment (saghul) [inline src/app.ts:79]: This statement is unnecessary, if we want the default to be redis. Just keep the default and perhaps add a comment which says `// redis`

File: src/app.ts [hunk 3]
```diff
@@ -48,28 +51,68 @@ const redisOptions = <RedisOptions>{
...
+
+switch (config.InstanceStoreProvider) {
+    // case 'consul':
+    //     instanceStore = new ConsulClient({
+    //         logger,
+    //         endpoint: config.ConsulURL,
+    //     });
+    //     break;
+    default:
+    case 'redis':
```
Comment (saghul) [inline src/app.ts:103]: ditto.

**Review by saghul — COMMENTED** [review]
I'm not very familiar with this code, but since you requested some eyes... here are some comments.

**Review by aaronkvanmeerten — COMMENTED** [review]
> why use a beta release? anything specific we need from it? 

This one supports the latest redis clients, including ioredis which we prefer.

**Review by saghul — APPROVED** [review]

**Pull request merged**